extension/chrome/content/download.js

```diff
@@ -83,18 +85,38 @@ Downloader = {
 
       // If target doesn't exist, create it
       if (!this.target.exists()) {
-        this.target.create(0x00, 0644);
+        this.target.create(Ci.nsIFile.NORMAL_FILE_TYPE, parseInt("0666", 8));
       }
 
-      this._persist = Cc["@mozilla.org/embedding/browser/nsWebBrowserPersist;1"].
-                      createInstance(Ci.nsIWebBrowserPersist);
-      this._persist.persistFlags = Ci.nsIWebBrowserPersist.PERSIST_FLAGS_REPLACE_EXISTING_FILES |
-                                   Ci.nsIWebBrowserPersist.PERSIST_FLAGS_FROM_CACHE |
-                                   Ci.nsIWebBrowserPersist.PERSIST_FLAGS_CLEANUP_ON_FAILURE;
-
-      // Save URL as target
-      this._persist.progressListener = new DownloaderListener();
-      this._persist.saveURI(uri, null, null, null, null, this.target, null);
+      Downloads.createDownload({'source': uri, 'target': this.target}).then((aDownload) => {
+        var maxMB = -1;
+        var currentMB = -1;
+        aDownload.onchange = function () {
+          if (aDownload.succeeded) {
+            var bytes = aDownload.hasProgress ?
+                        aDownload.totalBytes : aDownload.currentBytes;
+            maxMB = bytes / 1024 / 1024;
+          } else if (aDownload.hasProgress) {
+            // If the final size and progress are known, use them.
+            maxMB = aDownload.totalBytes / 1024 / 1024;
+            currentMB = aDownload.currentBytes / 1024 / 1024;
+          } else {
+            // The download final size and progress percentage is unknown.
+            Downloader.progressMeter.mode =  "undetermined";
+          }
+            var string = Downloader.stringBundle.getFormattedString("download.progress", [currentMB, maxMB]);
+            Downloader.progressLabel.value = string;
+
+            // Update progress meter values
+            Downloader.progressMeter.max = maxMB;
+            Downloader.progressMeter.value = currentMB;
+        };
+        aDownload.start();
+        aDownload.whenSucceeded().then(() => {
+           Downloader.progressLabel.value = Downloader.stringBundle.getString("download.finished");
+           Downloader.stop();
```

Comment:
Or maybe better we move the remaining updates for the status succeeded from above down to here, so it's in a single place.